src/foam/u2/crunch/wizardflow/GraphWizardletsAgent.js

```diff
@@ -0,0 +1,132 @@
+    'wizardlets'
+  ],
+
+  requires: [
+    'foam.nanos.crunch.ui.PrerequisiteAwareWizardlet',
+    'foam.u2.wizard.ProxyWAO'
+  ],
+
+  properties: [
+    {
+      name: 'wizardlets',
+      class: 'FObjectArray',
+      of: 'foam.u2.wizard.Wizardlet'
+    },
+    {
+      class: 'Map',
+      name: 'capabilityWizardletsMap'
+    }
+  ],
+
+  methods: [
+    async function execute() {
+      const prerequisites = [...this.capabilities];
+      const desiredCapability = prerequisites.pop();
+      this.wizardlets = await this.parseArrayToWizardlets(prerequisites, desiredCapability);
+
+      // Wire wizardlets into a DAG by populating 'prerequisiteWizardlets'
+      for ( 
+        const [capabilityId, {wizardlets}]
+        of Object.entries(this.capabilityWizardletsMap)
+      ) {
+        const graphNode = this.capabilityGraph.data[capabilityId];
+        const prerequisiteWizardlets = graphNode.forwardLinks
+          // Only add prerequisite wizardlets for capabilities with wizardlets
+          .filter(id => this.capabilityWizardletsMap[id])
+          // Always link the "primary" wizardlet of a capability
+          .map(id => this.capabilityWizardletsMap[id].primaryWizardlet)
+          ;
+
+        // Link the primary wizardlet of each prerequisite of this capability
+        //   to all of this capability's wizardlets
+        for ( let wizardlet of wizardlets ) {
+          wizardlet.prerequisiteWizardlets = prerequisiteWizardlets;
+        }
+      }
+    },
+    async function parseArrayToWizardlets(prereqs, parent) {
+      const wizardlets = [];
+
+      let x = { capability: parent };
+      const afterWizardlet = this.adaptWizardlet(x, parent.wizardlet);
+      const beforeWizardlet = this.adaptWizardlet(x, parent.beforeWizardlet);
+
+      const rootWizardlets = [beforeWizardlet, afterWizardlet].filter(exists => exists);
+
+      // Update a map of all capability IDs to their respective wizardlets
+      if ( ! this.capabilityWizardletsMap[parent.id] ) {
+        this.capabilityWizardletsMap[parent.id] = {
+          primaryWizardlet: afterWizardlet,
+          wizardlets: []
+        };
+      }
+      this.capabilityWizardletsMap[parent.id].wizardlets.push(...rootWizardlets);
+
+      if ( beforeWizardlet ) {
+        beforeWizardlet.isAvailable$.follow(afterWizardlet.isAvailable$);
+        afterWizardlet.data$ = beforeWizardlet.data$;
+      }
+
+      for ( let capability of prereqs ) {
+        // Perform a recursive call to get wizardlets for this prerequisite
+        const subPrereqs = Array.isArray(capability) ? [...capability] : [capability];
+        const subParent = subPrereqs.pop();
+        const subWizardlets = await this.parseArrayToWizardlets(subPrereqs, subParent);
+        const wizardlet = subWizardlets.pop();
+
+        // Call addPrerequisite if this prerequisite relation is effective
+        for ( let rootWizardlet of rootWizardlets ) {
+          if ( this.isPrerequisiteAware(rootWizardlet) ) {
```

Comment:
So BaseWizardlet is just has prerequisiteWizardlets and it's not prerequisiteAware. So this is saying add prerequisite to the wizardlet for the cases where we have prerequisiteAware wizardlets such as the MinMaxCapabilityWizardlet?

We'll always be storing prequisiteWizardlets regardless on any wizardlets too. So the only difference again with **PrerequisiteAwareWizardlet** and **BaseWizardlet** now is that an wizardlet that implements PrerequisiteAwareWizardlet would have an addPrerequisite method, and that method could detail how the wizardlets isAvailables slot together, etc?

Reply:
That seems to be the difference right now. The reason `PrerequisiteAware` was ever a thing is because the wizard didn't have knowledge of all the prerequisite relationships between the capabilities it's working with, but we're changing that. If we can eliminate the need for `PrerequisiteAware` on UCJ wizards then `addPrerequisite` might become a thing on `BaseWizardlet` that can be overridden, but for now its behaviour needs to remain intact or it might cause dozens of regressions.